Sources/SourceKitLSP/Swift/MacroExpansion.swift

```diff
@@ -0,0 +1,118 @@
+  /// and then makes `ShowDocumentRequest` to the client side for each
+  /// expansion to be displayed.
+  ///
+  /// - Parameters:
+  ///   - expandMacroCommand: The `ExpandMacroCommand` that triggered this request.
+  ///
+  /// - Returns: An `[RefactoringEdit]` with the necessary edits and buffer name as a `LSPAny`
+  func expandMacro(
+    _ expandMacroCommand: ExpandMacroCommand
+  ) async throws -> LSPAny {
+    guard let sourceKitLSPServer else {
+      // `SourceKitLSPServer` has been destructed. We are tearing down the
+      // language server. Nothing left to do.
+      throw ResponseError.unknown("Connection to the editor closed")
+    }
+
+    guard let sourceFileURL = expandMacroCommand.textDocument.uri.fileURL else {
+      throw ResponseError.unknown("Given URI is not a file URL")
+    }
+
+    let expansion = try await self.refactoring(expandMacroCommand)
+
+    for macroEdit in expansion.edits {
+      // buffer name without ".swift"
+      let macroExpansionBufferDirectoryName =
+        macroEdit.bufferName.hasSuffix(".swift")
+        ? String(macroEdit.bufferName.dropLast(6))
+        : macroEdit.bufferName
+
+      let macroExpansionBufferDirectoryURL = self.generatedMacroExpansionsPath
+        .appendingPathComponent(macroExpansionBufferDirectoryName)
+      do {
+        try FileManager.default.createDirectory(at: macroExpansionBufferDirectoryURL, withIntermediateDirectories: true)
+      } catch {
+        throw ResponseError.unknown(
+          "Failed to create directory for Macro Expansion Buffer at path: \(macroExpansionBufferDirectoryURL.path)"
+        )
+      }
+
+      let macroExpansionFileName = sourceFileURL.deletingPathExtension().lastPathComponent  // name of the source file
+      let macroExpansionPositionRangeIndicator =
+        "L\(macroEdit.range.lowerBound.line)C\(macroEdit.range.lowerBound.utf16index)-L\(macroEdit.range.upperBound.line)C\(macroEdit.range.upperBound.utf16index)"  // github permalink notation for position range
+
+      let macroExpansionFilePath = macroExpansionBufferDirectoryURL.appendingPathComponent(
+        "\(macroExpansionFileName)_\(macroExpansionPositionRangeIndicator).\(sourceFileURL.pathExtension)"
+      )
+
+      do {
+        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: .utf8)
+      } catch {
+        throw ResponseError.unknown("Unable to write macro expansion to file path: \"\(macroExpansionFilePath.path)\"")
+      }
+
+      Task {
+        let req = ShowDocumentRequest(uri: DocumentURI(macroExpansionFilePath), selection: macroEdit.range)
+
+        let response = await orLog("Sending ShowDocumentRequest to Client") {
+          try await sourceKitLSPServer.sendRequestToClient(req)
+        }
+
+        if let response, !response.success {
+          logger.error("client refused to show document for \(expansion.title, privacy: .public)")
+        }
+      }
+    }
+
+    return expansion.edits.encodeToLSPAny()
```

Comment:
Do we even need to return any edits here? I would expect that we have already handled the showing of the macro expansion buffers and the client shouldn’t need to deal with the edits returned from the `ExecuteCommandRequest`, right? Also, since it’s `LSPAny`, I would expect that the client doesn’t even know how to interpret it. Or am I missing something?

Edit: I now see that the tests rely on `ExecuteCommand` returning the edits. If we don’t need `expandMacro` and `semanticRefactoring` to return any results for editors, I think we should change this in a follow-up PR to get this one merged.

Reply:
The way the semantic refactoring for inline macro command was implemented is such that the testing part utilises the result. Since it was already accepted into the codebase, I thought it was a convention and implemented macro expansion command in a similar manner.

I guess it would be better to merge this PR and make a fix for this is in a follow-up PR as you said.

Reply:
Filed https://github.com/apple/sourcekit-lsp/issues/1498 so we don’t forget about it.